lib/datalog/access.go

```diff
@@ -0,0 +1,380 @@
+type EDB map[string][]Predicate
+
+// IDB (intensional database) type holds the interpreted facts from rules.
+type IDB map[string][]Predicate
+
+// NodeAccessRequest defines a request for access for a specific user, login, and node.
+type NodeAccessRequest struct {
+	Username  string
+	Login     string
+	Node      string
+	Namespace string
+}
+
+// AccessResponse defines all interpreted facts from rules.
+type AccessResponse struct {
+	Accesses        IDB
+	facts           EDB
+	mappings        map[string]uint32
+	reverseMappings map[uint32]string
+}
+
+const (
+	loginTraitHash = 0
+	keyJoin        = "_"
+
+	hasRole             = "has_role"
+	hasTrait            = "has_trait"
+	roleAllowsLogin     = "role_allows_login"
+	roleDeniesLogin     = "role_denies_login"
+	roleAllowsNodeLabel = "role_allows_node_label"
+	roleDeniesNodeLabel = "role_denies_node_label"
+	nodeHasLabel        = "node_has_label"
+
+	accesses     = "accesses"
+	denyAccesses = "deny_accesses"
+	denyLogins   = "deny_logins"
+
+	denyNullString   = "No denied accesses found.\n"
+	accessNullString = "No accesses found.\n"
+
+	userIndex  = 0
+	loginIndex = 1
+	nodeIndex  = 2
+	roleIndex  = 3
+)
+
+// QueryAccess returns a list of accesses to Teleport.
+func (c *NodeAccessRequest) QueryAccess(client auth.ClientI) (*AccessResponse, error) {
+	resp := AccessResponse{make(IDB), make(EDB), make(map[string]uint32), make(map[uint32]string)}
+	ctx := context.TODO()
+	resp.addToMap(types.Wildcard)
+
+	if c.Username != "" {
+		u, err := client.GetUser(c.Username, false)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		resp.createUserPredicates(u, c.Login)
+	} else {
+		us, err := client.GetUsers(false)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		for _, u := range us {
+			resp.createUserPredicates(u, c.Login)
+		}
+	}
+
+	rs, err := client.GetRoles(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	for _, r := range rs {
+		resp.createRolePredicates(r)
+	}
+
+	ns, err := client.GetNodes(ctx, c.Namespace)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	for _, n := range ns {
+		if len(c.Node) == 0 || n.GetHostname() == c.Node {
+			resp.createNodePredicates(n)
+		}
+	}
+
+	b, err := json.Marshal(resp.facts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	err = json.Unmarshal([]byte(C.GoString(C.process_access(C.CString(string(b))))), &resp.Accesses)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	resp.Accesses = cleanOutput(resp.Accesses, resp.reverseMappings)
+	resp.Accesses = filterByLogin(resp.Accesses, resp.reverseMappings, c.Login)
+	return &resp, nil
+}
+
+func (r *AccessResponse) addPredicate(key string, atoms ...interface{}) {
+	var mappedAtoms []uint32
+	for _, atom := range atoms {
+		id, ok := atom.(int)
+		if ok {
+			mappedAtoms = append(mappedAtoms, uint32(id))
+			continue
+		}
+		val, ok := atom.(string)
+		if !ok {
+			// Invalid type, we can just skip this predicate.
```

Comment:
If string is invalid atom type, can the caller just not pass it to this function?

Basically, it looks like we're trying to use type assertions here instead of relying on Go strong type system which isn't a proper usage of interfaces (IIUC what's going on here).